plugins/action_modify_record/src/databaseMethods.ts

```diff
@@ -1,56 +1,121 @@
+  createTable: async (tableName: string) => {
+    const text = `CREATE TABLE "${tableName}" ( id serial PRIMARY KEY)`
+    console.log('creating table with statement: ', text)
     try {
-      const result = await DBConnect.query({ text, values: Object.values(record) })
-      return { success: true, [tableName]: result.rows[0] }
+      await DBConnect.query({
+        text,
+        value: [tableName],
+      })
     } catch (err) {
       console.log(err.message)
-      throw err
+      throw new Error(`Filed to create table: ${tableName}`)
+    }
+  },
+  createJoinTableAndRecord: async (tableName: string, applicationId: number, recordId: number) => {
+    // Create join table if one doesn't exist
+    const joinTableName = `application_${tableName}_join`
+    const text = `CREATE TABLE IF NOT EXISTS ${joinTableName}( 
+                  id serial PRIMARY KEY, 
+                  application_id integer references application(id),
+                  ${tableName}_id integer references "${tableName}"(id));`
+
+    console.log('creating join table with statement: ', text)
+    try {
+      await DBConnect.query({ text })
+    } catch (err) {
+      console.log(err.message)
+      throw new Error(`Filed to create join table: ${joinTableName}`)
+    }
+    // Create entry in the joiin table
+    const joinTableRecord = { application_id: applicationId, [`${tableName}_id`]: recordId }
+    await createRecord(joinTableName, joinTableRecord, DBConnect)
+  },
+  createFields: async (
+    tableName: string,
+    fieldsToCreate: { fieldName: string; fieldType: string }[]
+  ) => {
+    const getType = (fieldType: string) => (fieldType === 'object' ? 'jsonb' : 'varchar')
+    const newColumns = fieldsToCreate.map(
+      ({ fieldName, fieldType }) => `ADD COLUMN "${fieldName}" ${getType(fieldType)}`
+    )
+    const text = `ALTER TABLE "${tableName}" ${newColumns.join(',')};`
+    console.log('creating new columns with statement: ', text)
+    try {
+      await DBConnect.query({
+        text,
+        value: [tableName],
+      })
+    } catch (err) {
+      console.log(err.message)
+      throw new Error(
+        `Filed to create fields in table table: ${tableName} ${JSON.stringify(
+          fieldsToCreate,
+          null,
+          ' '
+        )}`
+      )
     }
   },
 })
 
+const createRecord = async (
+  tableName: string,
+  record: { [key: string]: any },
+  DBConnect: DBConnectType
+) => {
+  const text = `
+    INSERT INTO "${tableName}" (${Object.keys(record)}) 
+    VALUES (${DBConnect.getValuesPlaceholders(record)})
+    RETURNING *
+    `
+  try {
+    const result = await DBConnect.query({ text, values: Object.values(record) })
+    const firstRow = result.rows[0]
+    return { success: true, [tableName]: firstRow, recordId: firstRow.id }
+  } catch (err) {
+    console.log(err.message)
+    throw err
+  }
+}
+
+export type DatabaseMethodsType = ReturnType<typeof databaseMethods>
```

Comment:
Since all Actions have (or will have) a databaseMethods import, should this type be defined in `/plugins/types` and shared to all Actions? (No need to adjust existing actions, but it'd be good to have it available)

Reply:
This one is specific to modifyRecord action.

Since i've added a method `createOrUpdateTable` to the action, which takes action specific instance of databaseMethods, needed to know what type it is.

Also I've noticed that `DBConnect?: any` is the parameter of ActionPluginInput, we should probably use the `DBConnectType` now. I didn't want to bloat this PR though, incase there are problems with types after that